internal/check/engine.go

```diff
@@ -65,7 +68,17 @@ func (e *Engine) CheckIsMember(ctx context.Context, r *relationTuple, restDepth
 // CheckRelationTuple checks if the relation tuple's subject has the relation on
 // the object in the namespace either directly or indirectly and returns a check
 // result.
-func (e *Engine) CheckRelationTuple(ctx context.Context, r *relationTuple, restDepth int) checkgroup.Result {
+func (e *Engine) CheckRelationTuple(ctx context.Context, r *relationTuple, restDepth int) (res checkgroup.Result) {
+	ctx, span := trace.SpanFromContext(ctx).TracerProvider().Tracer("keto/internal/check").Start(ctx, "Engine.CheckRelationTuple")
+	defer func() {
+		if res.Err != nil {
+			span.SetStatus(codes.Error, res.Err.Error())
+		} else {
+			span.SetAttributes(attribute.String("membership", res.Membership.String()))
+		}
+		span.End()
+	}()
+
```

Comment:
Since the function has only two returns, why do we need a defer for this? You could also do:

```
span.SetAttributes(attribute.String("membership", res.Membership.String()))
return result

err := ...
span.SetStatus(codes.Error, res.Err.Error())
return err
```

Reply:
There are two possible error returns from the function:
- `<-ctx.Done()`
- `<-resultCh` returns a struct with the Err field set.

Checking the named return value in the deferred call catches them both. Without that, the branch (`if res.Err != nil`) would need to move inside the `select` which is awkward.

Reply:
Not sure if I understand - would this not just be:

```patch
	resultCh := make(chan checkgroup.Result)
	go e.checkIsAllowed(ctx, r, restDepth)(ctx, resultCh)
	select {
	case result := <-resultCh:
+              span.SetAttributes(attribute.String("membership", res.Membership.String()))
		return result
	case <-ctx.Done():
+               span.SetStatus(codes.Error, ctx.Err())
		return checkgroup.Result{Err: errors.WithStack(ctx.Err())}
	}
```

Reply:
in the top branch, if `result.Err` is not `nil`, we would miss that error. Not sure if that can ever happen 🤷

Reply:
@aeneasr do you still want me to make this change? Otherwise this PR is ready to merge.

Reply:
I think it's a bit clonky to read compared to the two-liner, but it does guard against future regressions. It's also just a style choice so I'm fine either way - you decide :)